src/main/java/org/truffleruby/language/methods/LiteralMethodDefinitionNode.java

```diff
@@ -0,0 +1,86 @@
+
+import com.oracle.truffle.api.CallTarget;
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.object.DynamicObject;
+
+/**
+ * Define a method from a method literal (def mymethod ... end).
+ * That is, create an InternalMethod and add it to the current module (default definee).
+ */
+public class LiteralMethodDefinitionNode extends RubyNode {
+
+    private final String name;
+    private final SharedMethodInfo sharedMethodInfo;
+    private final CallTarget callTarget;
+    private final boolean isDefSingleton;
+
+    @Child private GetDefaultDefineeNode getDefaultDefineeNode;
+    @Child private RubyNode moduleNode;
+    @Child private RubyNode visibilityNode;
+
+    @Child private AddMethodNode addMethodNode;
+
+    public LiteralMethodDefinitionNode(RubyNode moduleNode, String name, SharedMethodInfo sharedMethodInfo, CallTarget callTarget, boolean isDefSingleton) {
+        this.name = name;
+        this.sharedMethodInfo = sharedMethodInfo;
+        this.callTarget = callTarget;
+        this.isDefSingleton = isDefSingleton;
+        this.moduleNode = moduleNode;
+        if (!isDefSingleton) {
+            this.visibilityNode = new GetCurrentVisibilityNode();
+        }
+        this.addMethodNode = AddMethodNodeGen.create(isDefSingleton, null, null, null);
+    }
+
+    @Override
+    public Object execute(VirtualFrame frame) {
+        final DynamicObject module = (DynamicObject) moduleNode.execute(frame);
+
+        final Visibility visibility;
+        if (isDefSingleton) {
+            visibility = Visibility.PUBLIC;
+        } else {
+            visibility = (Visibility) visibilityNode.execute(frame);
+        }
+
+        final DeclarationContext declarationContext = RubyArguments.getDeclarationContext(frame);
+        final InternalMethod currentMethod = RubyArguments.getMethod(frame);
+        final Object self = RubyArguments.getSelf(frame);
+        final DynamicObject capturedDefaultDefinee = captureDefaultDefinee(declarationContext, currentMethod, self);
+
+        final InternalMethod method = new InternalMethod(getContext(),
+                sharedMethodInfo, currentMethod.getLexicalScope(), name, module, visibility, false, null, callTarget, null, capturedDefaultDefinee);
+
+        addMethodNode.executeAddMethod(module, method, visibility);
+
+        return getSymbol(name);
+    }
+
+    @TruffleBoundary
+    private DynamicObject captureDefaultDefinee(DeclarationContext declarationContext, InternalMethod currentMethod, Object self) {
+        if (declarationContext == DeclarationContext.INSTANCE_EVAL) {
+            if (getDefaultDefineeNode == null) {
+                getDefaultDefineeNode = insert(new GetDefaultDefineeNode());
+            }
+            return getDefaultDefineeNode.getDefaultDefinee(currentMethod, self, declarationContext);
+        } else {
+            return null;
+        }
+    }
+
+}
```

Comment:
Does Git realise that this is a rename?

Reply:
It's not a rename, it's logic from 2 different files.